src/Compilers/Shared/GlobalAssemblyCacheHelpers/GacFileResolver.cs

```diff
@@ -20,7 +20,9 @@ internal sealed class GacFileResolver : IEquatable<GacFileResolver>
         /// <summary>
         /// Returns true if GAC is available on the current platform.
         /// </summary>
-        public static bool IsAvailable => typeof(object).Assembly.GlobalAssemblyCache;
+        public static bool IsAvailable
+            // Check a type from System.Core since Mono doesn't load mscorlib from the GAC.
```

Comment:
This type is actually in `System.Linq.dll` in certain distributions. Part of the problem is that it's hard to depend on the assembly location of a given type because the runtime team moves them around. 

Think a better approach would be to pick a type like this then assert it's in a different assembly than `object`. That way we catch ourselves if this ever gets invalidated.

Reply:
@jaredpar  Any idea why mscorlib is not loaded from GAC?

Reply:
Hmm.. I guess what I'm really looking for is an Assembly that will always load from the GAC for platforms that support a GAC and a repeatable method for locating the Assembly. mscorlib does not satisfy that requirement for at least some versions of Mono.

Could look through `AppDomain.GetAssemblies()` to see if any were loaded from the GAC and cache the result.

Reply:
That doesn't seem like it would be reliable in the same way checking a specific one isn't. Maybe we could detect presence of GAC on Mono in a different way.

Reply:
> Any idea why mscorlib is not loaded from GAC?

On mono? No idea.

> Maybe we could detect presence of GAC on Mono in a different way.

Is there a case on Mono where the GAC isn't used?

Reply:
Yes, on Mono. The current code works fine on .NET Framework. 

> Is there a case on Mono where the GAC isn't used?

Apparently there is - @JoeRobich just found out.

Maybe @marek-safar knows?

Reply:
For Mono GAC is optional (there is also option to have multiple GACs). The most reliable way is probably check GAC location calling into https://github.com/mono/mono/blob/master/mcs/class/corlib/System/Environment.cs#L975 and then compare the paths/MVIDs/what-ever-is-appropriate

Reply:
How does the runtime know to look into GAC? Does it always look there and if it's not found it looks somewhere else? Where else?

Reply:
There is whole assembly loader logic with tons of rules and config options (it's similar to FW assembly binder). There is no documentation covering all the possible combinations and the behaviour though. You'd need to read the code.

Reply:
@JoeRobich I think the best thing we can do here to return true if we run on Mono. http://sourceroslyn.io/#Microsoft.CodeAnalysis.Scripting/Hosting/Resolvers/MonoGlobalAssemblyCache.cs,160 that will get called if this property returns true.